src/main/java/org/jenkinsci/plugins/envinject/EnvInjectPlugin.java

```diff
@@ -0,0 +1,143 @@
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.envinject;
+
+import hudson.Plugin;
+import hudson.XmlFile;
+import hudson.model.Descriptor;
+import hudson.model.Run;
+import hudson.security.Permission;
+import hudson.security.PermissionGroup;
+import hudson.security.PermissionScope;
+import java.io.File;
+import java.io.IOException;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import javax.servlet.ServletException;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.export.ExportedBean;
+
+/**
+ * @author Oleg Nenashev
+ */
+@ExportedBean
+public class EnvInjectPlugin extends Plugin {
+
+    private final static Logger LOGGER = Logger.getLogger(EnvInjectPlugin.class.getName());
+
+    public static final PermissionGroup PERMISSIONS = new PermissionGroup(EnvInjectPlugin.class, Messages._envinject_permissions_title());
+
+    /**
+     * Allows to view injected variables.
+     * Even Jenkins admins may have no such permission in particular installations.
+     */
+    public static final Permission VIEW_INJECTED_VARS = new Permission(PERMISSIONS, "ViewVars", Messages._envinject_permissions_viewVars_description(), null, PermissionScope.RUN);
+     
+    private EnvInjectPluginConfiguration configuration;
+
+    /**
+     * Retrieves the plugin instance.
+     * @return {@link EnvInjectPlugin}
+     * @throws IllegalStateException the plugin has not been loaded yet
+     */
+    public static @Nonnull EnvInjectPlugin getInstance() {
+        Jenkins j = Jenkins.getInstance();
+        EnvInjectPlugin plugin = j != null ? j.getPlugin(EnvInjectPlugin.class) : null;
+        if (plugin == null) { // Fail horribly
+            // TODO: throw a graceful error
+            throw new IllegalStateException("Cannot get the plugin's instance. Jenkins or the plugin have not been initialized yet");
+        }
+        return plugin;
+    }
+    
+    @Override
+    protected void load() throws IOException {
+        super.load();
+        if (configuration == null) {     
+            configuration = EnvInjectPluginConfiguration.getDefault();        
+            save();
+        }
+    }
+
+    void configure(EnvInjectPluginConfiguration configuration) throws IOException {
+        this.configuration = configuration;
+        VIEW_INJECTED_VARS.setEnabled(configuration.isEnablePermissions());
```

Comment:
@oleg-nenashev agreed, which is why I think a proper solution to JENKINS-17200 is important. Not sure about the design yet. Anyway out of scope here.